### PR TITLE
Facing and Axis to enum

### LINKED ISSUE #89: Type hinting PHPStan for Facing
It may be useful to specify the expected return type of Facing functions with: `value-of<Facing::ALL>`.  

This can break many static analyses, because depending on the PHPstan levels, it will carry over if you pass a simple integer into a function accepting just a few values.  

I think this option is worth considering, as it could prevent certain runtime errors in the event of an unaccepted value being supplied?

### PR DESCRIPTION
**Pull request by DebHugo**
Fix #89

This change requires many changes in pmmp. I can do if this pr is good for you.

### CODE REVIEW

File: src/Facing.php [hunk 1]
```diff
@@ -56,6 +56,9 @@ private function __construct(){
 		self::EAST
 	];
 
+	/**
+	 * @deprecated
+	 */
 	public const OFFSET = [
```
Comment (dktapps) [inline src/Facing.php]: why is this marked as deprecated?

Comment (DebHugo) [reply src/Facing.php]: It's marked as obsolete because i'm replacing it with a function. In my opinion, it is clearer.

With constant:
```
Facing::OFFSET[Facing::DOWN->value]
```

With function:
```
Facing::DOWN->offset()
```

Comment (dktapps) [reply src/Facing.php]: This seems outside of the PR scope

File: src/Facing.php [hunk 2]
```diff
@@ -89,67 +95,60 @@ private function __construct(){
...
-		return ($direction & self::FLAG_AXIS_POSITIVE) === self::FLAG_AXIS_POSITIVE;
+	public static function isPositive(Facing $direction) : bool{
+		return ($direction->value & self::FLAG_AXIS_POSITIVE) === self::FLAG_AXIS_POSITIVE;
 	}
 
 	/**
 	 * Returns the opposite Facing of the specified one.
 	 *
-	 * @param int $direction 0-5 one of the Facing::* constants
+	 * @throws \ValueError if opposite facing don't exist
```
Comment (dktapps) [inline src/Facing.php]: This is impossible

**Comment by dktapps** [comment]
If we're going with enums I would prefer to ditch the numeric values entirely. They have no purpose except to be implicitly misused as Minecraft constants.

**Comment by DebHugo** [comment]
> If we're going with enums I would prefer to ditch the numeric values entirely. They have no purpose except to be implicitly misused as Minecraft constants.

Should I do the same for Axis? I don't know if these values ​​represent anything or not

**Comment by dktapps** [comment]
> > If we're going with enums I would prefer to ditch the numeric values entirely. They have no purpose except to be implicitly misused as Minecraft constants.
> 
> Should I do the same for Axis? I don't know if these values ​​represent anything or not

Yes

File: src/Facing.php [hunk 3]
```diff
@@ -89,67 +88,58 @@ private function __construct(){
...
 	 * Returns the opposite Facing of the specified one.
-	 *
-	 * @param int $direction 0-5 one of the Facing::* constants
 	 */
-	public static function opposite(int $direction) : int{
-		return $direction ^ self::FLAG_AXIS_POSITIVE;
+	public static function opposite(Facing $direction) : Facing{
+		return match ($direction) {
+			self::DOWN => self::UP,
+			self::EAST => self::SOUTH,
```
Comment (dktapps) [inline src/Facing.php]: This is incorrect

File: src/Facing.php [hunk 4]
```diff
@@ -89,67 +88,58 @@ private function __construct(){
...
+			self::WEST => self::EAST,
+			self::UP => self::DOWN,
+			self::SOUTH => self::EAST,
```
Comment (dktapps) [inline src/Facing.php]: also incorrect

File: src/Facing.php [hunk 5]
```diff
@@ -89,67 +88,58 @@ private function __construct(){
...
+		};
 	}
 
 	/**
 	 * Returns whether the direction is facing the positive of its axis.
 	 */
-	public static function isPositive(int $direction) : bool{
-		return ($direction & self::FLAG_AXIS_POSITIVE) === self::FLAG_AXIS_POSITIVE;
+	public static function isPositive(Facing $direction) : bool{
+		return in_array($direction, [self::UP, self::SOUTH, self::EAST]);
```
Comment (dktapps) [inline src/Facing.php]: `in_array()` is an immediate red flag whenever I see it

File: src/Facing.php [hunk 6]
```diff
@@ -89,67 +88,58 @@ private function __construct(){
...
-	public static function opposite(int $direction) : int{
-		return $direction ^ self::FLAG_AXIS_POSITIVE;
+	public static function opposite(Facing $direction) : Facing{
+		return match ($direction) {
+			self::DOWN => self::UP,
+			self::NORTH => self::SOUTH,
+			self::WEST => self::EAST,
+			self::UP => self::DOWN,
+			self::SOUTH => self::NORTH,
+			self::EAST => self::WEST,
```
Comment (dktapps) [inline src/Facing.php:86]: This should follow the order of the cases - it's a confusing mess right now

File: src/Facing.php [hunk 7]
```diff
@@ -89,67 +88,58 @@ private function __construct(){
...
-		if(!isset(self::CLOCKWISE[$axis])){
-			throw new \InvalidArgumentException("Invalid axis $axis");
-		}
-		if(!isset(self::CLOCKWISE[$axis][$direction])){
-			throw new \InvalidArgumentException("Cannot rotate facing \"" . self::toString($direction) . "\" around axis \"" . Axis::toString($axis) . "\"");
-		}
-
-		$rotated = self::CLOCKWISE[$axis][$direction];
+	public static function rotate(Facing $direction, Axis $axis, bool $clockwise) : Facing{
+		$rotated = self::CLOCKWISE[$axis->value][$direction->value];
```
Comment (dktapps) [inline src/Facing.php]: These don't have values anymore

File: src/Facing.php [hunk 8]
```diff
@@ -159,17 +149,17 @@ public static function validate(int $facing) : void{
 	}
```
Comment (dktapps) [inline src/Facing.php:139]: There's no point deprecating this. We're making sweeping BC breaks here anyway.

File: src/Facing.php [hunk 9]
```diff
@@ -159,17 +149,17 @@ public static function validate(int $facing) : void{
...
 
 	/**
+	 * @deprecated use Facing->name
```
Comment (dktapps) [inline src/Facing.php]: If this is a valid replacement, why isn't the function using it?

File: src/Facing.php (hunk 6 above)
Comment (Dhaiven) [reply src/Facing.php:116]: How i can do ?

File: src/Facing.php (hunk 8 above)
Comment (DebHugo) [reply src/Facing.php:139]: So i remove this function ?

File: src/Facing.php (hunk 9 above)
Comment (DebHugo) [reply src/Facing.php]: Indead

File: src/AxisAlignedBB.php [hunk 10]
```diff
@@ -136,20 +136,18 @@ public function offsetCopy(float $x, float $y, float $z) : AxisAlignedBB{
...
 	 * Offsets this AxisAlignedBB in the given direction by the specified distance.
 	 *
-	 * @param int $face one of the Facing::* constants
-	 *
 	 * @return $this
 	 */
-	public function offsetTowards(int $face, float $distance) : AxisAlignedBB{
-		[$offsetX, $offsetY, $offsetZ] = Facing::OFFSET[$face] ?? throw new \InvalidArgumentException("Invalid Facing $face");
+	public function offsetTowards(Facing $face, float $distance) : AxisAlignedBB{
+		[$offsetX, $offsetY, $offsetZ] = Facing::OFFSET[$face->value];
```
Comment (DebHugo) [inline src/AxisAlignedBB.php]: Since faces don't have values, this creates a bug. To fix this bug, why not introduce an offset() function like this:
```
public function offset(): array {
		return match($this){
			self::DOWN  => [ 0, -1,  0],
			self::UP    => [ 0, +1,  0],
			self::NORTH => [ 0,  0, -1],
			self::SOUTH => [ 0,  0, +1],
			self::WEST  => [-1,  0,  0],
			self::EAST  => [+1,  0,  0]
		};
	}

#Result
Face::DOWN->offset()
```

I already tried to add this previously and you say it's outside of this pr.
I agree but I don't see any other solutions to solve this problem.
I await your opinion.

Comment (dktapps) [reply src/AxisAlignedBB.php]: Yeah, I don't see another way to do it without bringing back the values.

**Comment by DebHugo** [comment]
This pr is good ? if yes i can change pocketmine to support Facing enum

File: src/Facing.php [hunk 11]
```diff
@@ -56,120 +38,92 @@ private function __construct(){
...
-			self::SOUTH => self::UP
-		]
-	];
-
 	/**
 	 * Returns the axis of the given direction.
 	 */
-	public static function axis(int $direction) : int{
-		return $direction >> 1; //shift off positive/negative bit
+	public static function axis(Facing $direction) : Axis{
```
Comment (dktapps) [inline src/Facing.php]: Doesn't make sense for this to still be static

File: src/Facing.php [hunk 12]
```diff
@@ -56,120 +38,92 @@ private function __construct(){
...
+			self::EAST  => [+1,  0,  0]
+		};
 	}
 
 	/**
 	 * Returns whether the direction is facing the positive of its axis.
 	 */
-	public static function isPositive(int $direction) : bool{
-		return ($direction & self::FLAG_AXIS_POSITIVE) === self::FLAG_AXIS_POSITIVE;
+	public static function isPositive(Facing $direction) : bool{
```
Comment (dktapps) [inline src/Facing.php]: As above

File: src/Facing.php [hunk 13]
```diff
@@ -56,120 +38,92 @@ private function __construct(){
...
 	}
 
 	/**
 	 * Returns the opposite Facing of the specified one.
-	 *
-	 * @param int $direction 0-5 one of the Facing::* constants
 	 */
-	public static function opposite(int $direction) : int{
-		return $direction ^ self::FLAG_AXIS_POSITIVE;
+	public static function opposite(Facing $direction) : Facing{
```
Comment (dktapps) [inline src/Facing.php]: As above

File: src/Facing.php [hunk 14]
```diff
@@ -56,120 +38,92 @@ private function __construct(){
...
-	 * @throws \InvalidArgumentException if not possible to rotate $direction around $axis
 	 */
-	public static function rotate(int $direction, int $axis, bool $clockwise) : int{
-		if(!isset(self::CLOCKWISE[$axis])){
-			throw new \InvalidArgumentException("Invalid axis $axis");
-		}
-		if(!isset(self::CLOCKWISE[$axis][$direction])){
-			throw new \InvalidArgumentException("Cannot rotate facing \"" . self::toString($direction) . "\" around axis \"" . Axis::toString($axis) . "\"");
-		}
+	public static function rotate(Facing $direction, Axis $axis, bool $clockwise) : Facing{
```
Comment (dktapps) [inline src/Facing.php]: As above

File: src/Facing.php [hunk 15]
```diff
@@ -56,120 +38,92 @@ private function __construct(){
...
 
-		$rotated = self::CLOCKWISE[$axis][$direction];
 		return $clockwise ? $rotated : self::opposite($rotated);
 	}
 
-	/**
-	 * @throws \InvalidArgumentException
-	 */
-	public static function rotateY(int $direction, bool $clockwise) : int{
+	public static function rotateY(Facing $direction, bool $clockwise) : Facing{
```
Comment (dktapps) [inline src/Facing.php]: As above

File: src/Facing.php [hunk 16]
```diff
@@ -56,120 +38,92 @@ private function __construct(){
...
 		return self::rotate($direction, Axis::Y, $clockwise);
 	}
 
-	/**
-	 * @throws \InvalidArgumentException
-	 */
-	public static function rotateZ(int $direction, bool $clockwise) : int{
+	public static function rotateZ(Facing $direction, bool $clockwise) : Facing{
```
Comment (dktapps) [inline src/Facing.php]: As above

File: src/Facing.php [hunk 17]
```diff
@@ -56,120 +38,92 @@ private function __construct(){
...
 		return self::rotate($direction, Axis::Z, $clockwise);
 	}
 
-	/**
-	 * @throws \InvalidArgumentException
-	 */
-	public static function rotateX(int $direction, bool $clockwise) : int{
+	public static function rotateX(Facing $direction, bool $clockwise) : Facing{
```
Comment (dktapps) [inline src/Facing.php]: As above

File: src/Vector3.php [hunk 18]
```diff
@@ -170,19 +170,19 @@ public function east(int $step = 1){
...
-	 * @return \Generator|Vector3[]
-	 * @phpstan-return \Generator<int, Vector3, void, void>
+	 * @return \Generator|mixed[] [Facing $facing, Vector3 $vector]
+	 * @phpstan-return \Generator<int, array{Facing, Vector3}, void, void>
 	 */
 	public function sides(int $step = 1) : \Generator{
-		foreach(Facing::ALL as $facing){
-			yield $facing => $this->getSide($facing, $step);
+		foreach(Facing::cases() as $facing){
+			yield [$facing, $this->getSide($facing, $step)];
```
Comment (dktapps) [inline src/Vector3.php]: This didn't need to be changed. Generators allow yielding objects as the key.

File: src/Vector3.php [hunk 19]
```diff
@@ -170,19 +170,19 @@ public function east(int $step = 1){
...
 		}
 	}
 
 	/**
 	 * Same as sides() but returns a pre-populated array instead of Generator.
 	 *
-	 * @return Vector3[]
+	 * @return array<int, array{Facing,Vector3}>
```
Comment (dktapps) [inline src/Vector3.php]: Same here

File: src/Vector3.php [hunk 20]
```diff
@@ -191,15 +191,13 @@ public function sidesArray(bool $keys = false, int $step = 1) : array{
...
+	 * @return \Generator|mixed[] [Facing $facing, Vector3 $vector]
+	 * @phpstan-return \Generator<int, array{Facing, Vector3}, void, void>
 	 */
-	public function sidesAroundAxis(int $axis, int $step = 1) : \Generator{
-		foreach(Facing::ALL as $facing){
+	public function sidesAroundAxis(Axis $axis, int $step = 1) : \Generator{
+		foreach(Facing::cases() as $facing){
 			if(Facing::axis($facing) !== $axis){
-				yield $facing => $this->getSide($facing, $step);
+				yield [$facing, $this->getSide($facing, $step)];
```
Comment (dktapps) [inline src/Vector3.php]: Same here

**Review by dktapps — CHANGES_REQUESTED** [review]

File: src/Facing.php [hunk 21]
```diff
@@ -56,120 +38,92 @@ private function __construct(){
 		self::EAST
 	];
```
Comment (DebHugo) [inline src/Facing.php:39]: To keep the new logic, i think `HORIZONTAL` constant must be removed and replace by 2 functions:
- `horizontal(): array` -> return same value likes `HORIZONTAL` constant
- `isHorizontal(): bool` -> return true if the face is horizontal else false
I await your opinion

Comment (dktapps) [reply src/Facing.php:39]: I don't think that's necessary. This is only used for iteration anyway.

File: src/Facing.php [hunk 22]
```diff
@@ -56,120 +38,92 @@ private function __construct(){
...
+		return match ($this) {
+			self::DOWN, self::UP => Axis::Y,
+			self::NORTH, self::SOUTH => Axis::Z,
+			self::WEST, self::EAST => Axis::X,
+		};
 	}
 
 	/**
-	 * Returns whether the direction is facing the positive of its axis.
+	 * @return non-empty-array<int>
```
Comment (dktapps) [inline src/Facing.php]: PHPStan doc comments should be prefixed by @phpstan-

File: src/AxisAlignedBB.php [hunk 23]
```diff
@@ -465,20 +448,23 @@ public function calculateIntercept(Vector3 $pos1, Vector3 $pos2) : ?RayTraceResu
...
+			[Facing::UP, $v4],
+			[Facing::NORTH, $v5],
+			[Facing::SOUTH, $v6]
+		] as $value){
+			$v = $value[1];
 			if($v !== null and ($d = $pos1->distanceSquared($v)) < $distance){
 				$vector = $v;
 				$distance = $d;
-				$face = $f;
+				/** @var Facing $face */
```
Comment (dktapps) [inline src/AxisAlignedBB.php]: Don't use @var tags like this, they'll suppress legit errors

File: src/AxisAlignedBB.php [hunk 24]
```diff
@@ -465,20 +448,23 @@ public function calculateIntercept(Vector3 $pos1, Vector3 $pos2) : ?RayTraceResu
 
 		$vector = null;
 		$distance = PHP_INT_MAX;
-		$face = -1;
+		/** @var Facing $face */
```
Comment (dktapps) [inline src/AxisAlignedBB.php]: Same here

**Review by dktapps — CHANGES_REQUESTED** [review]

File: src/Facing.php [hunk 25]
```diff
@@ -56,120 +38,112 @@ private function __construct(){
...
+	public function axis() : Axis{
+		return match ($this) {
+			self::DOWN, self::UP => Axis::Y,
+			self::NORTH, self::SOUTH => Axis::Z,
+			self::WEST, self::EAST => Axis::X,
+		};
+	}
+
+	/**
+	 * @phpstan-return non-empty-array<int>
```
Comment (dktapps) [inline src/Facing.php]:
```suggestion
	 * @phpstan-return array{-1|0|1, -1|0|1, -1|0|1}
```

File: src/Facing.php [hunk 26]
```diff
@@ -56,120 +38,112 @@ private function __construct(){
...
-			throw new \InvalidArgumentException("Cannot rotate facing \"" . self::toString($direction) . "\" around axis \"" . Axis::toString($axis) . "\"");
-		}
+	public function rotate(Axis $axis, bool $clockwise) : Facing{
+		$rotated = match ($axis) {
+			Axis::Y => match ($this) {
+				self::NORTH => self::EAST,
+				self::EAST => self::SOUTH,
+				self::SOUTH => self::WEST,
+				self::WEST => self::NORTH,
+				default => throw new \InvalidArgumentException("Face " . strtolower($this->name) . " not match with Axis " . strtolower($axis->name))
```
Comment (dktapps) [inline src/Facing.php]: These error messages should match the originals

File: src/Facing.php [hunk 27]
```diff
@@ -56,120 +38,112 @@ private function __construct(){
...
+			},
+			Axis::Z => match ($this) {
+				self::UP => self::EAST,
+				self::EAST => self::DOWN,
+				self::DOWN => self::WEST,
+				self::WEST => self::UP,
+				default => throw new \InvalidArgumentException("Face " . strtolower($this->name) . " not match with Axis " . strtolower($axis->name))
```
Comment (dktapps) [inline src/Facing.php]: and here

File: src/Facing.php [hunk 28]
```diff
@@ -56,120 +38,112 @@ private function __construct(){
...
+			},
+			Axis::X => match ($this) {
+				self::UP => self::NORTH,
+				self::NORTH => self::DOWN,
+				self::DOWN => self::SOUTH,
+				self::SOUTH => self::UP,
+				default => throw new \InvalidArgumentException("Face " . strtolower($this->name) . " not match with Axis " . strtolower($axis->name))
```
Comment (dktapps) [inline src/Facing.php]: and here

File: tests/phpunit/FacingTest.php [hunk 29]
```diff
@@ -35,9 +35,9 @@ public static function axisProvider() : \Generator{
 		yield [Facing::EAST, Axis::X];
 	}
 
-	#[DataProvider("axisProvider")]
-	public function testAxis(int $direction, int $axis) : void{
-		self::assertEquals($axis, Facing::axis($direction));
+  #[DataProvider("axisProvider")]
```
Comment (dktapps) [inline tests/phpunit/FacingTest.php]: Unrelated formatting change

File: src/Facing.php [hunk 30]
```diff
@@ -56,120 +38,112 @@ private function __construct(){
...
-	public static function rotateZ(int $direction, bool $clockwise) : int{
-		return self::rotate($direction, Axis::Z, $clockwise);
+	public function rotateZ(bool $clockwise) : Facing{
+		return $this->rotate(Axis::Z, $clockwise);
 	}
 
 	/**
+	 * Rotates the given direction around the X axis.
+	 * 
+	 * @see Facing::rotate()
```
Comment (dktapps) [inline src/Facing.php]: Unrelated change

File: src/Facing.php [hunk 31]
```diff
@@ -56,120 +38,112 @@ private function __construct(){
...
-	public static function rotateY(int $direction, bool $clockwise) : int{
-		return self::rotate($direction, Axis::Y, $clockwise);
+	public function rotateY(bool $clockwise) : Facing{
+		return $this->rotate(Axis::Y, $clockwise);
 	}
 
 	/**
+	 * Rotates the given direction around the Z axis.
+	 * 
+	 * @see Facing::rotate()
```
Comment (dktapps) [inline src/Facing.php]: Unrelated

File: src/Facing.php [hunk 32]
```diff
@@ -56,120 +38,112 @@ private function __construct(){
...
 
-		$rotated = self::CLOCKWISE[$axis][$direction];
-		return $clockwise ? $rotated : self::opposite($rotated);
+		return $clockwise ? $rotated : $rotated->opposite();
 	}
 
 	/**
+	 * Rotates the given direction around the Y axis.
+	 * 
+	 * @see Facing::rotate()
```
Comment (dktapps) [inline src/Facing.php]: Unrelated

File: src/Facing.php [hunk 33]
```diff
@@ -56,120 +38,112 @@ private function __construct(){
...
+		};
 	}
 
 	/**
 	 * Returns whether the direction is facing the positive of its axis.
 	 */
-	public static function isPositive(int $direction) : bool{
-		return ($direction & self::FLAG_AXIS_POSITIVE) === self::FLAG_AXIS_POSITIVE;
+	public function isPositive() : bool{
+		return $this === self::UP || $this === self::SOUTH || $this === self::EAST;
```
Comment (dktapps) [inline src/Facing.php]: I'd prefer if this used a match with all cases.
It doesn't affect anything in this particular case, but it's generally better hygiene for enums, since it'll guarantee an error will appear if more cases are added.

File: src/Facing.php [hunk 34]
```diff
@@ -56,120 +38,112 @@ private function __construct(){
...
-		]
-	];
-
 	/**
 	 * Returns the axis of the given direction.
 	 */
-	public static function axis(int $direction) : int{
-		return $direction >> 1; //shift off positive/negative bit
+	public function axis() : Axis{
+		return match ($this) {
```
Comment (dktapps) [inline src/Facing.php]:
```suggestion
		return match($this){
```

File: src/Facing.php [hunk 35]
```diff
@@ -56,120 +38,112 @@ private function __construct(){
...
+			self::NORTH, self::SOUTH => Axis::Z,
+			self::WEST, self::EAST => Axis::X,
+		};
+	}
+
+	/**
+	 * @phpstan-return non-empty-array<int>
+	 */
+	public function offset() : array{
+		return match ($this) {
```
Comment (dktapps) [inline src/Facing.php]:
```suggestion
		return match($this){
```

File: src/Facing.php [hunk 36]
```diff
@@ -56,120 +38,112 @@ private function __construct(){
...
 
 	/**
 	 * Returns the opposite Facing of the specified one.
-	 *
-	 * @param int $direction 0-5 one of the Facing::* constants
 	 */
-	public static function opposite(int $direction) : int{
-		return $direction ^ self::FLAG_AXIS_POSITIVE;
+	public function opposite() : Facing{
+		return match ($this) {
```
Comment (dktapps) [inline src/Facing.php]:
```suggestion
		return match($this){
```

File: src/Facing.php [hunk 37]
```diff
@@ -56,120 +38,112 @@ private function __construct(){
...
+			self::WEST => self::EAST,
+			self::EAST => self::WEST,
+		};
 	}
 
 	/**
 	 * Rotates the given direction around the axis.
 	 *
-	 * @throws \InvalidArgumentException if not possible to rotate $direction around $axis
+	 * @throws \InvalidArgumentException
```
Comment (dktapps) [inline src/Facing.php]: Unrelated change

File: src/Facing.php [hunk 38]
```diff
@@ -56,120 +38,112 @@ private function __construct(){
...
 	 */
-	public static function rotate(int $direction, int $axis, bool $clockwise) : int{
-		if(!isset(self::CLOCKWISE[$axis])){
-			throw new \InvalidArgumentException("Invalid axis $axis");
-		}
-		if(!isset(self::CLOCKWISE[$axis][$direction])){
-			throw new \InvalidArgumentException("Cannot rotate facing \"" . self::toString($direction) . "\" around axis \"" . Axis::toString($axis) . "\"");
-		}
+	public function rotate(Axis $axis, bool $clockwise) : Facing{
+		$rotated = match ($axis) {
```
Comment (dktapps) [inline src/Facing.php]:
```suggestion
		$rotated = match($axis){
```

**Review by dktapps — CHANGES_REQUESTED** [review]

File: src/Facing.php (hunk 30 above)
Comment (DebHugo) [reply src/Facing.php]: I keep the @throws \InvalidArgumentException or not ?

File: src/Facing.php (hunk 33 above)
Comment (DebHugo) [reply src/Facing.php]: Like this
```
return match($this){
    self::UP, self::SOUTH, self::EAST => true,
    self::DOWN, self::NORTH, self::WEST => false,
    default => throw new \InvalidArgumentException("Cannot know if facing \"" . strtolower($this->name) . "\"is positive or not")
};
```
or witth `default => false` ?

File: src/Facing.php (hunk 30 above)
Comment (dktapps) [reply src/Facing.php]: Just remove the stuff that was added that isn't related to the PR

File: src/Facing.php [hunk 39]
```diff
@@ -56,120 +38,112 @@ private function __construct(){
...
+				self::WEST => self::UP,
+				default => throw new \InvalidArgumentException("Cannot rotate facing \"" . strtolower($this->name) . "\" around axis \"" . strtolower($axis->name) . "\"")
+			},
+			Axis::X => match($this){
+				self::UP => self::NORTH,
+				self::NORTH => self::DOWN,
+				self::DOWN => self::SOUTH,
+				self::SOUTH => self::UP,
+				default => throw new \InvalidArgumentException("Cannot rotate facing \"" . strtolower($this->name) . "\" around axis \"" . strtolower($axis->name) . "\"")
+			}
```
Comment (dktapps) [inline src/Facing.php:117]: The default cases could just fall through to `null`, then we could just check if `$rotated` is `null` after the match so as not to duplicate exception messages 3 times

File: src/Facing.php (hunk 33 above)
Comment (dktapps) [reply src/Facing.php]: Like that, but delete the `default` case
